src/noise_interfaces/virtual_brownian_tree_interface.jl

```diff
@@ -0,0 +1,325 @@
+
+  if Z0 != nothing
+    Zs = [Z0, Zend]
+  else
+    Zs = [nothing]
+  end
+
+  seeds = [convert(typeof(rng.ctr1), rng.ctr1+(Nt+1)/2)]
+
+  for level in 1:Int(tree_depth)
+    new_ts = Vector{typeof(t0)}(undef, 0)
+    new_Ws = Vector{typeof(W0)}(undef, 0)
+    if Z0 != nothing
+      new_Zs = Vector{typeof(Z0)}(undef, 0)
+    else
+      new_Zs = [nothing]
+    end
+    new_seeds = Vector{typeof(rng.ctr1)}(undef, 0)
+    for (i, parent) in enumerate(seeds)
+      seed_l, seed_r, seed_v = split_VBT_seed(rng,parent,level,Nt)
+      append!(new_seeds,[seed_l, seed_r])
+
+      t0, t1 = ts[i], ts[i+1]
+      W0, W1 = Ws[i], Ws[i+1]
+      t = (t0 + t1) / 2
+
+      h = t1-t0
+      q = (t-t0)/h # == 1//2
+
+      w = bridge(dW,nothing,W0,W1,q,h,nothing,nothing,nothing,rng)
+      if Z0 != nothing
+        Z0, Z1 = Zs[i], Zs[i+1]
+        z = bridge(dW,nothing,Z0,Z1,q,h,nothing,nothing,nothing,rng)
+        append!(new_Zs,Z0)
+        append!(new_Zs,z)
+      end
+
+      append!(new_ts,[t0,t])
+      append!(new_Ws,[W0,w])
+    end
+    push!(new_ts,tend)
+    push!(new_Ws,Wend)
+
+    if Z0 != nothing
+      push!(new_Zs,Zend)
+      Zs = new_Zs
+    end
+
+    ts = new_ts
+    Ws = new_Ws
+
+    seeds = new_seeds
+  end
+
+  return ts, Ws, Zs, seeds
+end
+
+
+function search_VBT(t,seed,t0,t1,W0,W1,Z0,Z1,W::VirtualBrownianTree,rng::Random123.AbstractR123)
+
+  Nt = Int(2^W.search_depth+1)
+  depth = Int(W.tree_depth+1)
+  seed_l, seed_r, seed_v = split_VBT_seed(rng,seed,depth,Nt)
+
+  tmid = (t0 + t1) / 2
+  h = t1-t0
+  q = 1//2
+
+  out1 = W.bridge(W.dW,nothing,W0,W1,q,h,nothing,nothing,nothing,rng)
+
+  if Z0 != nothing
+    out2 = W.bridge(W.dW,nothing,Z0,Z1,q,h,nothing,nothing,nothing,rng)
+  else
+    out2 = nothing
+  end
+
+  # create a buffer
+  W0tmp, W1tmp = copy(W0), copy(W1)
+  if Z0 != nothing
+    Z0tmp, Z1tmp = copy(Z0), copy(Z1)
+  end
+
+  while (abs(t-tmid) > W.atol && depth<W.search_depth)
+    depth += 1
+    if t < tmid
+      t1 = tmid
+      W0tmp, W1tmp = W0tmp, out1
+      seed_v = seed_l
+      if Z0 != nothing
+        Z0tmp, Z1tmp = Z0tmp, out2
+      end
+    else
+      t0 = tmid
+      W0tmp, W1tmp = out1, W1tmp
+      seed_v = seed_r
+      if Z0 != nothing
+        Z0tmp, Z1tmp = out2, Z1tmp
+      end
+    end
+
+    seed_l, seed_r, seed_v = split_VBT_seed(rng,seed_v,depth,Nt)
+    tmid = (t0 + t1) / 2
+    h = t1-t0
+
+    out1 = W.bridge(W.dW,nothing,W0tmp,W1tmp,q,h,nothing,nothing,nothing,rng)
+    if Z0 != nothing
+      out2 = W.bridge(W.dW,nothing,Z0tmp,Z1tmp,q,h,nothing,nothing,nothing,rng)
+    end
+  end
+
+  return out1, out2
+end
+
+
+function search_VBT!(out1,out2,t,seed,t0,t1,W0,W1,Z0,Z1,W::VirtualBrownianTree,rng::Random123.AbstractR123)
+
+  Nt = Int(2^W.search_depth+1)
+  depth = Int(W.tree_depth+1)
+  seed_l, seed_r, seed_v = split_VBT_seed(rng,seed,depth,Nt)
+
+  tmid = (t0 + t1) / 2
+  h = t1-t0
+  q = 1//2
+
+  # create a buffer
+  W0tmp, W1tmp = copy(W0), copy(W1)
```

Comment:
should these tmp values be moved to the VBT struct (to make it fully non-allocating in the `search_VBT` call)?

Reply:
I think so. Keeping them out of the struct doesn't decrease the peak memory usage anyways, so decreasing the allocations could be a good call.